### PR TITLE
MaterialLimitTimeStep postprocessor in SM and TM

### PR DESCRIPTION
**Pull request by acasagran**
Reimplementation of MaterialLimitTimeStep postprocessor in SM and TM for IsotropicPlasticity and PowerLawCreep material models. Uses the effective plastic or creep strain scalar quantity to compute the maximum time step size based on the user supplied maximum strain increment. Added regression tests for each model demonstrating the new postprocessor. Refs #8553

### CODE REVIEW

File: modules/solid_mechanics/src/materials/ReturnMappingModel.C [hunk 1]
```diff
@@ -186,3 +190,17 @@ ReturnMappingModel::computeStress(const Elem & /*current_elem*/,
...
+
+Real
+ReturnMappingModel::computeTimeStepLimit()
+{
+  Real tmp_max_incr;
+  Real scalar_inelastic_strain_incr;
+  unsigned int qp;
+
+  scalar_inelastic_strain_incr =
+      _effective_inelastic_strain[qp] - _effective_inelastic_strain_old[qp];
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/ReturnMappingModel.C]: You never assign a value to `qp`!!!

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/solid_mechanics/include/materials/CombinedCreepPlasticity.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ class CombinedCreepPlasticity : public ConstitutiveModel
   bool _output_iteration_info;
   Real _relative_tolerance;
   Real _absolute_tolerance;
+  MaterialProperty<Real> & _matl_timestep_limit;
 
 private:
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/CombinedCreepPlasticity.h]: remove empty access qualifier

File: modules/solid_mechanics/include/materials/ReturnMappingModel.h [hunk 3]
```diff
@@ -63,8 +65,10 @@ class ReturnMappingModel : public ConstitutiveModel
 
   MaterialProperty<Real> & _effective_inelastic_strain;
   MaterialProperty<Real> & _effective_inelastic_strain_old;
+  Real _max_inelastic_incr;
 
 private:
+  Real _max_inc;
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/ReturnMappingModel.h]: why does this need to be private?

File: modules/solid_mechanics/src/materials/CombinedCreepPlasticity.C [hunk 4]
```diff
@@ -168,6 +169,13 @@ CombinedCreepPlasticity::computeStress(const Elem & current_elem,
   }
 
   strain_increment = elastic_strain_increment;
+
+  _matl_timestep_limit[qp] = 0.0;
+  for (unsigned i_rmm(0); i_rmm < num_submodels; ++i_rmm)
+  {
+    _matl_timestep_limit[qp] += 1.0 / rmm[i_rmm]->computeTimeStepLimit();
+  }
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/CombinedCreepPlasticity.C]: remove curly braces

File: modules/solid_mechanics/src/materials/CombinedCreepPlasticity.C [hunk 5]
```diff
@@ -168,6 +169,13 @@ CombinedCreepPlasticity::computeStress(const Elem & current_elem,
   }
 
   strain_increment = elastic_strain_increment;
+
+  _matl_timestep_limit[qp] = 0.0;
+  for (unsigned i_rmm(0); i_rmm < num_submodels; ++i_rmm)
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/CombinedCreepPlasticity.C]: `i_rmm = 0` !

File: modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h [hunk 6]
```diff
@@ -68,6 +69,10 @@ class RadialReturnStressUpdate : public StressUpdateBase
 
   MaterialProperty<Real> & _effective_inelastic_strain;
   MaterialProperty<Real> & _effective_inelastic_strain_old;
+  Real _max_inelastic_incr;
+
+private:
+  Real _max_inc;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h]: why private?

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C [hunk 7]
```diff
@@ -280,6 +281,9 @@ ComputeMultipleInelasticStress::updateQpStateSingleModel(
                          combined_inelastic_strain_increment,
                          _Jacobian_mult[_qp]);
 
+  _matl_timestep_limit[_qp] = 0.0;
+  _matl_timestep_limit[_qp] = _models[0]->computeTimeStepLimit();
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C]: is the first assignment necessary?

File: modules/tensor_mechanics/src/postprocessors/MaterialTimeStepPostprocessor.C [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+void
+MaterialTimeStepPostprocessor::execute()
+{
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    computeQpValue();
+}
+
+void
+MaterialTimeStepPostprocessor::computeQpValue()
```
Comment (dschwen) [inline modules/tensor_mechanics/src/postprocessors/MaterialTimeStepPostprocessor.C]: Factoring this method out looks a bit overengineered.

File: modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h (hunk 6 above)
Comment (permcody) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h]: Just because I see these comments often, I want to say that I disagree that protected should be the default accessor mode in the MOOSE (and the modules). Private accessors ensure object integrity which goes out the window once you go protected. WIth a code like MOOSE where every object is inherited, there's nothing wrong with considering which objects belong in private and which belong in protected. Protected with an inheritance heavy code gives many objects free reign to mess with base class state where it may be difficult to understand all of the corner case ways that an object's integrity can be compromised.

Comment (dschwen) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h]: `private` is absolutely correct and necessary _if_ object integrity needs to be ensured. But in most cases I encounter there is _no_ integrity at stake. It seems to me that this member should just be protected const. This is member is not representing an internal state that needs to be set or read using a particular accessor.

Comment (permcody) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h]: `protected` with `const` is always a 👍 . 
However many times it's not. As long as we are on the same page. It's always a fine line between making the framework accessible and protecting against corruption

File: modules/solid_mechanics/src/materials/CombinedCreepPlasticity.C (hunk 5 above)
Comment (dschwen) [inline modules/solid_mechanics/src/materials/CombinedCreepPlasticity.C]: ok, this is in solid mechanics, I guess I don't care then

**Review by dschwen — APPROVED** [review]

File: modules/solid_mechanics/src/materials/ReturnMappingModel.C [hunk 9]
```diff
@@ -42,7 +44,9 @@ ReturnMappingModel::ReturnMappingModel(const InputParameters & parameters,
     _effective_inelastic_strain(
         declareProperty<Real>("effective_" + inelastic_strain_name + "_strain")),
     _effective_inelastic_strain_old(
-        declarePropertyOld<Real>("effective_" + inelastic_strain_name + "_strain"))
+        declarePropertyOld<Real>("effective_" + inelastic_strain_name + "_strain")),
+    _max_inelastic_incr(isParamValid("max_inelastic_incr") ? getParam<Real>("max_inelastic_incr")
+                                                           : 0.)
```
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/ReturnMappingModel.C]: I'm coming into this late since this has been merged, but I stumbled on this code, and see a few issues.  First off, this use of isParamValid doesn't really make sense. It looks like it's setting it to 0 if it's not set, but then that parameter has a default value of 1e-4, so it would always use that as the default.

File: modules/solid_mechanics/src/materials/ReturnMappingModel.C [hunk 10]
```diff
@@ -26,6 +26,8 @@ validParams<ReturnMappingModel>()
       "relative_tolerance", 1e-5, "Relative convergence tolerance for sub-newtion iteration");
   params.addParam<Real>(
       "absolute_tolerance", 1e-20, "Absolute convergence tolerance for sub-newtion iteration");
+  params.addParam<Real>(
+      "max_inelastic_incr", 1e-4, "The maximum inelastic strain increment allowed in a time step");
```
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/ReturnMappingModel.C]: I really don't like this name.  If we spared a few characters and made this `max_inelastic_increment`, it would be more consistent with our other parameters.

File: modules/solid_mechanics/src/materials/ReturnMappingModel.C [hunk 11]
```diff
@@ -186,3 +190,16 @@ ReturnMappingModel::computeStress(const Elem & /*current_elem*/,
...
+Real
+ReturnMappingModel::computeTimeStepLimit(unsigned qp)
+{
+  Real tmp_max_incr;
+  Real scalar_inelastic_strain_incr;
+
+  scalar_inelastic_strain_incr =
+      _effective_inelastic_strain[qp] - _effective_inelastic_strain_old[qp];
+  tmp_max_incr = _max_inelastic_incr * 1e-10;
+  _max_inc = std::max(scalar_inelastic_strain_incr, tmp_max_incr);
```
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/ReturnMappingModel.C]: I don't see any reason for `_max_inc` to be a member variable because this is the only place it is used. Just make it local to this method.  Also, I find this code hard to follow.  There's a magic factor of 1e-10 coming out of the blue, and _max_inc seems to be mixing the current strain increment and the allowable increment somehow. Is this just about avoiding a divide by zero?  If so, it would be better to just check if the denominator is close to zero (using `MooseUtils::absoluteFuzzyEqual(_max_inc, 0.0`), and return `std::numeric_limits<Real>::max()` if it is.

**Comment by bwspenc** [comment]
@acasagran I know this code has already been merged, but would you mind cleaning up these few minor issues I pointed out in a separate PR?
